File: python/system_manager_client.py
import logging
from system_manager_types import PowerState
import systemmanagerservice_pb2_grpc
import systemmanagerservice_signals_pb2
import systemmanagerservice_types_pb2
from time import sleep
from threading import Thread
import grpc

logger = logging.getLogger(__name__)

class SystemManagerClient:
    def __init__(self):
        self.channel = None
        self.stub = None
        self.running = False
        self.send_thread = Thread(target = self.send_keep_active)
        self.translation_to_grpc = {
            systemmanagerservice_types_pb2.POWERSTATE__ACTIVE: PowerState.ACTIVE,
            systemmanagerservice_types_pb2.POWERSTATE__TRANSITION_TO_ACTIVE: PowerState.TRANSITION_TO_ACTIVE,
            systemmanagerservice_types_pb2.POWERSTATE__LPM: PowerState.LPM,
            systemmanagerservice_types_pb2.POWERSTATE__TRANSITION_TO_LPM: PowerState.TRANSITION_TO_LPM
        }

    def connect(self):
        # self.channel = grpc.insecure_channel('localhost:50051', options=(('grpc.enable_http_proxy', 0),))
        self.channel = grpc.insecure_channel('localhost:50051')
        self.stub = (
            systemmanagerservice_pb2_grpc.SystemManagerServiceStub(
                self.channel
            )
        )
        self.running = True

    def disconnect(self):
        logger.info(f"Disconnect")
        self.channel = None
        self.stub = None
        self.running = False
        self.send_thread.join()

    def wait_power_state(self):
        logger.info(f"+ Wait for power state")
        return self.stub.NotifyOnPowerStateChange(
            systemmanagerservice_signals_pb2.NotifyOnPowerStateChangeRequest(client_id="test")
        )

    def start_sending_keep_active(self):
        self.send_thread.start()

    def send_keep_active(self):
        while self.running:
            logger.info(f"Request keep active")
            self.stub.NotifyOnKeepActive(self._generate_keep_request())
            sleep(1)

    def _generate_keep_request(self):
        yield systemmanagerservice_signals_pb2.KeepActiveRequest(client_id = "test")

    def acknowledge_power_state(self, powerStateEvent):
        logger.info(f"+ Ack power state: {powerStateEvent}")
        self.stub.AcknowledgePowerStateChange(
            systemmanagerservice_signals_pb2.AcknowledgePowerStateChangeRequest(
                client_id = "test", state = powerStateEvent.new_state))
        logger.info(f"+ Exit ack power state")

def run():
    logger.info(f"Enter run")
    client = SystemManagerClient()
    client.connect()

    client.start_sending_keep_active()

    exit_flag = False
    while (True):
        for response in client.wait_power_state():
            logger.info(f"Received power state {response}")
            client.acknowledge_power_state(response)
            if response.new_state == systemmanagerservice_types_pb2.POWERSTATE__ACTIVE:
                exit_flag = True
                break
        if exit_flag:
            break

    logger.info(f"LOGN BUTTON PRESS")
    sleep(5)
    client.disconnect()

if __name__ == '__main__':
    formatter = "[%(asctime)s] %(name)s {%(filename)s:%(lineno)d} %(levelname)s - %(message)s"
    logging.basicConfig(level = logging.INFO, format = formatter)
    run()

File: cpp/system_manager_client.cpp
#include <chrono>
#include <iostream>
#include <memory>
#include <random>
#include <string>
#include <thread>
#include <atomic>

#include <grpc/grpc.h>
#include <grpcpp/channel.h>
#include <grpcpp/client_context.h>
#include <grpcpp/create_channel.h>
#include <grpcpp/security/credentials.h>
#include "systemmanagerservice.grpc.pb.h"

using grpc::Channel;
using grpc::ClientContext;
using grpc::ClientReader;
using grpc::ClientReaderWriter;
using grpc::ClientWriter;
using grpc::Status;

namespace grpcexamples {
std::ostream& operator<<(std::ostream& os, PowerState state) {
    switch (state) {
        case POWERSTATE__OFF:
            os << "POWERSTATE__OFF";
            break;
        case POWERSTATE__TRANSITION_TO_LPM:
            os << "POWERSTATE__TRANSITION_TO_LPM";
            break;
        case POWERSTATE__LPM:
            os << "POWERSTATE__LPM";
            break;
        case POWERSTATE__TRANSITION_TO_ACTIVE:
            os << "POWERSTATE__TRANSITION_TO_ACTIVE";
            break;
        case POWERSTATE__ACTIVE:
            os << "POWERSTATE__ACTIVE";
            break;
    }
    return os;
}

class SystemManagerClient {
public:
    SystemManagerClient(std::shared_ptr<Channel> channel) :
        stub_(SystemManagerService::NewStub(channel)),
        m_running{true},
        m_threadSendKeepActive{std::thread(&SystemManagerClient::threadHandler, this)} {}

    void threadHandler() {
        std::cout << "Enter thread handler\n";
        while(m_running.load()) {
            sendKeepActive();
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
    }

    void sendKeepActive() {
        ClientContext context{};
        NoResponseRequired response{};
        std::unique_ptr< ClientWriter< KeepActiveRequest > > writer(
            stub_->NotifyOnKeepActive(&context, &response));

        KeepActiveRequest request{};
        request.set_client_id(CLIENT_ID);
        if (!writer->Write(request)) {
            std::cout << "Write failed!\n";
        }
        std::cout << "Send keep active success\n";
    }

    PowerState waitPowerState() {
        ClientContext context{};
        NotifyOnPowerStateChangeRequest request{};
        request.set_client_id(CLIENT_ID);
        std::unique_ptr<ClientReader<PowerStateChangeEvent>> reader(
            stub_->NotifyOnPowerStateChange(&context, request));

        PowerStateChangeEvent state{};        
        while(reader->Read(&state)) {
            std::cout << "Received notification for state " << state.new_state() << "\n";
        }

        Status status = reader->Finish();
        if (status.ok()) {
            std::cout << "Client was successfully notified\n" ;
        } else {
            std::cout << "Client notification failed\n" ;
        }

        return state.new_state();
    }

    void acknowledgePowerState(PowerState powerState) {
        ClientContext context{};
        AcknowledgePowerStateChangeRequest request{};
        request.set_client_id(CLIENT_ID); request.set_state(powerState);
        NoResponseRequired response{};
        Status status = stub_->AcknowledgePowerStateChange(&context, request, &response);
        std::cout << "Acnowledged state " << powerState << " for client " << CLIENT_ID << "\n";
    }

    ~SystemManagerClient() {
        m_running.store(false);
        m_threadSendKeepActive.join();
    }

private:
    std::unique_ptr<SystemManagerService::Stub> stub_;
    std::thread m_threadSendKeepActive;
    std::atomic_bool m_running{};
    const std::string CLIENT_ID = "cpp_client";
};
} // grpcexamples

int main(int argc, char** argv) {
    grpcexamples::SystemManagerClient client(
        grpc::CreateChannel("localhost:50051", grpc::InsecureChannelCredentials()));

    while(true) {
        const auto new_state = client.waitPowerState();
        client.acknowledgePowerState(new_state);
        if (grpcexamples::POWERSTATE__ACTIVE == new_state) {
            break;
        }
    }
    std::cout << "Exit main\n";
  
    return 0;
}
